content/blog/2023-12-11-rvv-llvm-gisel/index.md

```diff
@@ -0,0 +1,177 @@
+
+# What Was the Goal?
+The goal was to support LLVM Global Instruction Selection for RISC-V Vector Extensions on some ALU operations, such as `vadd`, `vsub`, `vand`, `vor`, and `vxor`.
+
+
+# A Primer on the RISC-V Vector Extension
+There are a lot of instructions and cool features in the [RISC-V "V" Vector Extension Spec](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc), I will only pick the most relevant and important parts to explain. 
+
+The most important parameters are undoubtly `SEW`, `VLEN`, `VL`, and `LMUL`; and one of the most interesting and powerful instructions is `vset{i}vl{i}`.
+
+Let's begin with the crucial parameters:
+- `SEW`: Selected Element Width (in bits), set dynamically by the programmers. It sets the width/length of a single element in a vector element. Each vector element can compose multiple single element.
+- `VLEN`: Vector register LENgth (in bits). The number of bits in a single vector register. It is hardware dependent. 
+- `VL`: Vector element Length (in bits) that the programmers actually deal with, which can be treated as the vector operation building blocks. It defines how many elements the vector operations will execute.
+- `LMUL`: The vector Length MULtiplier. It is used for grouping vector registers. It is a power of 2 and it ranges from 1/8 to 8. For instance, when `LMUL=8`, only `v0`, `v8`, `v16`, and `v24` indices are allowed to used, as for example, group `v8` encodes 8 vector elments `v8v9`...`v15`. Note it can also be fraction numbers because sometimes we want to use only parts of the vector registers.
+  
+After introducing these basic and the most important parameters, there are still two paramters we will be dealing with, `AVL` and `VLMAX`:
+- `AVL`: Application Vector Length. The application specifies the total number of elements to be processed as a candidate for `VL`.
+- `VLMAX`: MAX Vector Length. Although programmers can dynamically select vector register lengths, they cannot be set to arbitrary numbers because they might be larger than the hardware constraints, which is imposed by `VLMAX`. Specifically, `VLMAX=(LMUL × VLEN) / SEW`. For example, if the length of a single basic elment is 32 bit, and this specific hardware supports `VLEN`=256 bit, there are 8 basic blocks in a vector element. And we want to group every `LMUL`=2 vector elements, `VLMAX`=4.  As a result, the actual value `VL` is the smaller value of the desired application vector length `AVL` and the vector length value constrained by the hardware `VLMAX`: `VL = min(AVL, VLMAX)`.
```

Comment:
`8 basic blocks` might be confusing terminology since basic blocks usually refers to the sequence of instructions that have continuous control flow. Maybe it would be better to say `8 elements can fit in vector register`